### ISSUE TITLE
ArrayIndexOutOfBoundsException occurs in HlsChunkSource

### ISSUE THREAD

**Issue by Leeyoungryun** [google/ExoPlayer#5816]
Hi I'm using exoplayer version 2.9.0

In HlsChunkSource,

There is a code

`HlsMediaPlaylist.Segment segment = mediaPlaylist.segments.get(chunkIndex);`

however, if chunk index is negative, that chunkIndex cause ArrayIndexOutOfBoundsException.

`int chunkIndex = (int) (chunkMediaSequence - mediaPlaylist.mediaSequence);`

upper code, that chunkMediaSequence can be negative. if previous chunk's next chunk index is negative, that chunkMediaSequence is negative

I have a question, In what situations is previous chunk index to be negative? 

my log

```
2019-04-29 17:32:47.174 E/ExoPlayerImplInternal: Internal runtime error.
    java.lang.ArrayIndexOutOfBoundsException: length=10; index=-2
        at java.util.ArrayList.get(ArrayList.java:439)
        at java.util.Collections$UnmodifiableList.get(Collections.java:1351)
        at java.util.Collections$UnmodifiableList.get(Collections.java:1351)
        at com.google.android.exoplayer2.source.hls.HlsChunkSource.getNextChunk(HlsChunkSource.java:308)
        at com.google.android.exoplayer2.source.hls.HlsSampleStreamWrapper.continueLoading(HlsSampleStreamWrapper.java:561)
        at com.google.android.exoplayer2.source.CompositeSequenceableLoader.continueLoading(CompositeSequenceableLoader.java:78)
        at com.google.android.exoplayer2.source.hls.HlsMediaPeriod.continueLoading(HlsMediaPeriod.java:247)
        at com.google.android.exoplayer2.MediaPeriodHolder.continueLoading(MediaPeriodHolder.java:160)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.maybeContinueLoading(ExoPlayerImplInternal.java:1580)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleContinueLoadingRequested(ExoPlayerImplInternal.java:1552)
        at com.google.android.exoplayer2.ExoPlayerImplInternal.handleMessage(ExoPlayerImplInternal.java:324)
        at android.os.Handler.dispatchMessage(Handler.java:102)
        at android.os.Looper.loop(Looper.java:214)
        at android.os.HandlerThread.run(HandlerThread.java:65)

```

**Comment by Leeyoungryun** [comment]
I investigated some codes.

In HlsChunkSource

"
if (chunkMediaSequence < mediaPlaylist.mediaSequence) {
      if (previous != null && switchingVariant) {
        // We try getting the next chunk without adapting in case that's the reason for falling
        // behind the live window.
        selectedVariantIndex = oldVariantIndex;
        selectedUrl = variants[selectedVariantIndex];
        mediaPlaylist = playlistTracker.getPlaylistSnapshot(selectedUrl);
        startOfPlaylistInPeriodUs =
            mediaPlaylist.startTimeUs - playlistTracker.getInitialStartTimeUs();
        chunkMediaSequence = previous.getNextChunkIndex();
      } else {
        fatalError = new BehindLiveWindowException();
        return;
      }
    }
int chunkIndex = (int) (chunkMediaSequence - mediaPlaylist.mediaSequence);
"

I have a few questions about that code.

1. Do you consider if previous is not null and switchingVariant is true, chunkMediaSequence is under than  mediaPlaylist.mediaSequence?
 I saw many safe guard(queue is null then switchingVariant is almost false) for sinking with chunks, but upper codition often breaks out.

2. I think, "switchingVariant is true" means "Threre are some changes about track"  
is that right?

**Comment by AquilesCanta** [comment]
I think an index out of bounds may be thrown for live content in some scenarios. A fix should be available in the dev-branch relatively soon. However, I'm not sure the fix will make it to 2.10.

**Comment by ojw28** [comment]
The fix will be included in `2.10.0`; thanks!
